crates/ironrdp-rdpdr/src/pdu/efs.rs

```diff
@@ -2391,3 +2400,433 @@ impl ClientDriveQueryDirectoryResponse {
         }
     }
 }
+
+/// [2.2.3.3.6] Server Drive Query Volume Information Request
+///
+/// We only need to read the buffer up to the FileInformationClass to get the job done, so the rest of the fields in
+/// this structure are discarded. See FreeRDP:
+/// https://github.com/FreeRDP/FreeRDP/blob/511444a65e7aa2f537c5e531fa68157a50c1bd4d/channels/drive/client/drive_main.c#L464
+///
+/// [2.2.3.3.6]: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/484e622d-0e2b-423c-8461-7de38878effb
+#[derive(Debug)]
+pub struct ServerDriveQueryVolumeInformationRequest {
+    pub device_io_request: DeviceIoRequest,
+    pub fs_info_class_lvl: FileSystemInformationClassLevel,
+}
+
+impl ServerDriveQueryVolumeInformationRequest {
+    const NAME: &str = "DR_DRIVE_QUERY_VOLUME_INFORMATION_REQ";
+    pub fn decode(dev_io_req: DeviceIoRequest, src: &mut ReadCursor<'_>) -> PduResult<Self> {
+        ensure_size!(in: src, size: size_of::<u32>());
+        let fs_info_class_lvl = FileSystemInformationClassLevel::from(src.read_u32());
+
+        // This field MUST contain one of the following values.
+        static VALID_LEVELS: [FileSystemInformationClassLevel; 5] = [
+            FileSystemInformationClassLevel::FILE_FS_VOLUME_INFORMATION,
+            FileSystemInformationClassLevel::FILE_FS_SIZE_INFORMATION,
+            FileSystemInformationClassLevel::FILE_FS_ATTRIBUTE_INFORMATION,
+            FileSystemInformationClassLevel::FILE_FS_FULL_SIZE_INFORMATION,
+            FileSystemInformationClassLevel::FILE_FS_DEVICE_INFORMATION,
+        ];
+
+        if !VALID_LEVELS.contains(&fs_info_class_lvl) {
+            return Err(invalid_message_err!(
+                "ServerDriveQueryVolumeInformationRequest::decode",
+                "fs_info_class_lvl",
+                "received invalid level"
+            ));
+        }
+
+        // We only need to read the buffer up to the FileInformationClass to get the job done, so the rest of the fields in
+        // this structure are discarded. See FreeRDP:
+        // https://github.com/FreeRDP/FreeRDP/blob/511444a65e7aa2f537c5e531fa68157a50c1bd4d/channels/drive/client/drive_main.c#L464
+        ensure_size!(in: src, size: size_of::<u32>());
+        let length = cast_length!("ServerDriveQueryVolumeInformationRequest", "length", src.read_u32())?; // Length
+        ensure_size!(in: src, size: 24);
```

Comment:
Suggestion: you could define a `FIXED_PART_SIZE` const and ensure enough bytes for it using `ensure_fixed_part_size!`

Reply:
Good call, updated in 7f9ce0a942f7d982e9d7be05bbecd9dfec72e941